### PR TITLE
Properly no-op deprecations and allow unlisting at the same time in API

### PR DESCRIPTION
**Pull request by joelverhagen**
There is continued internal success and growing interesting in the deprecation API. I would like to make progress on the public preview.

Spec: https://github.com/NuGet/Engineering/pull/4709

For no-oping deprecations, I opted to not use the entity level change tracking API (e.g. https://stackoverflow.com/a/30235307) since it would mean a lot of changes to our abstractions. I think the current implementation is easier to understand anyway. 

Summary of changes:
- Properly no-op package deprecations that have no difference
   - https://github.com/NuGet/NuGetGallery/issues/9921
- Allow packages to be unlisted or relisted at the same time as deprecation
   - https://github.com/NuGet/NuGetGallery/issues/8873#issuecomment-2062110640
- Fix a problem where API deprecations were being audited as via the web UI (incorrect)
- Resolve some minor build warnings due to obsolete APIs and cases with no `await` in and `async` method
   - Unrelated but small

### CODE REVIEW

File: src/NuGetGallery/Services/IPackageDeprecationManagementService.cs [hunk 1]
```diff
@@ -22,12 +22,14 @@ public interface IPackageDeprecationManagementService
...
         /// <param name="id">The ID of the package to update.</param>
         /// <param name="versions">The versions of the package to update.</param>
+        /// <param name="auditReason">A reason description used for auditing purposes.</param>
         /// <param name="isLegacy">Whether or not the packages are legacy.</param>
         /// <param name="hasCriticalBugs">Whether or not the packages have critical bugs.</param>
         /// <param name="isOther">Whether or not the packages have an unlisted reason for being deprecated.</param>
         /// <param name="alternatePackageId">An alternate package ID to use instead.</param>
         /// <param name="alternatePackageVersion">A version of <paramref name="alternatePackageId"/> to use instead.</param>
         /// <param name="message">A custom message to add to the deprecation.</param>
+        /// <param name="listed">Null to leave current listed status per package version, false to unlist each package, true to relist the package.</param>
```
Comment (jimmylewis) [inline src/NuGetGallery/Services/IPackageDeprecationManagementService.cs:32]: nit: would this be more intuitive as an enum?
```
enum ListedVerb
{
    Unchanged,
    Unlist,
    Relist,
}
```

Comment (joelverhagen) [reply src/NuGetGallery/Services/IPackageDeprecationManagementService.cs:32]: Sounds good! I think it's unlikely for the tri-state to ever grow beyond that but even so this is clearer.

Comment (joelverhagen) [reply src/NuGetGallery/Services/IPackageDeprecationManagementService.cs]: Done.

File: src/NuGetGallery/Services/PackageDeprecationService.cs [hunk 2]
```diff
@@ -49,23 +50,36 @@ public class PackageDeprecationService : IPackageDeprecationService
                 throw new ArgumentException(nameof(packages));
             }
 
+            if (auditReason == null)
+            {
+                throw new ArgumentNullException(nameof(auditReason));
+            }
+
             var registration = packages.First().PackageRegistration;
```
Comment (sbanni) [inline src/NuGetGallery/Services/PackageDeprecationService.cs:58]: Unused?

Comment (joelverhagen) [reply src/NuGetGallery/Services/PackageDeprecationService.cs]: Fixed. Good catch!

**Review by sbanni — APPROVED** [review]

**Review by ryuyu — APPROVED** [review]

**Pull request merged**